signac/core/collection_api.py

```diff
@@ -0,0 +1,446 @@
+import os
+import json
+import errno
+import uuid
+from copy import copy
+from contextlib import contextmanager
+from collections.abc import Mapping
+from collections.abc import MutableMapping
+from collections.abc import Sequence
+from collections.abc import MutableSequence
+
+try:
+    import numpy
+    NUMPY = True
+except ImportError:
+    NUMPY = False
+
+try:
+    from collections.abc import Collection
+except ImportError:
+    # create Collection
+
+    from collections.abc import Sized, Iterable, Container
+
+    def _check_methods(C, *methods):
+        mro = C.__mro__
+        for method in methods:
+            for B in mro:
+                if method in B.__dict__:
+                    if B.__dict__[method] is None:
+                        return NotImplemented
+                    break
+            else:
+                return NotImplemented
+        return True
+
+    class Collection(Sized, Iterable, Container):  # type: ignore
+        @classmethod
+        def __subclasshook__(cls, C):
+            if cls is Collection:
+                return _check_methods(C,  "__len__", "__iter__", "__contains__")
+            return NotImplemented
+
+        @classmethod
+        def __instancecheck__(cls, instance):
+            for parent in cls.__mro__:
+                if not isinstance(instance, parent):
+                    return False
+            return True
+
+
+class SyncedCollection(Collection):
```

Comment:
Should this class be abstract? Should it have abstract methods like `to_base`?

Reply:
`SyncedCollection` is an abstract class as it inherits from `Collection` (`__contains__, __iter__, __len__` are not defined). I have not defined `load` and `sync` as abstract so that I can create an instance of `SyncedDict` and `SyncedList`. But not sure about `to_base`. Should I define it as an abstract method?

Reply:
You're right, it is abstract already. When do you need to instantiate `SyncedDict` or `SyncedList`? Should those classes (and `_load`, `_sync`) also be abstract, until you have a backend specified? I agree that `to_base` should be abstract.

Reply:
I agree, make methods abstract that must be implemented by specific back ends. In the long run, I don't think anything in the second level of the inheritance tree should be concrete, i.e. both SyncedDict and JSONCollection should be abstract. Until both parts are combined into a JSONDict, it shouldn't be possible to instantiate.